Graphics/Implicit/ExtOpenScad/Eval/Constant.hs

```diff
@@ -52,52 +51,29 @@ import Graphics.Implicit.ExtOpenScad.Parser.Lexer (matchTok)
 addConstants :: [String] -> IO (VarLookup, [Message])
 addConstants constants = do
   path <- getCurrentDirectory
-  let scadOpts = ScadOpts False False
-  (_, CompState (varLookup, _, _, messages, _)) <- liftIO $ runStateT (execAssignments constants 0) (CompState (defaultObjects, [], path, [], scadOpts))
-  pure (varLookup, messages)
+  (_, s) <- liftIO . runStateT (execAssignments constants) $ CompState defaultObjects [] path [] opts
+  pure (varLookup s, messages s)
   where
-    execAssignments :: [String] -> Fastℕ  -> StateC ()
-    execAssignments [] _ = pure ()
-    execAssignments (assignment:xs) count = do
-      let
-        pos = SourcePosition count 1 "cmdline_constants"
-        show' err = showErrorMessages "or" "unknown parse error" "expecting" "unexpected" "end of input" (errorMessages err)
-      case parseAssignment "cmdline_constant" assignment of
-        Left e            -> addMessage SyntaxError pos $ show' e
-        Right (key, expr) -> do
-          res <- evalExpr pos expr
-          case matchPat key res of
-            Nothing -> pure ()
-            Just pat -> modifyVarLookup $ varUnion pat
-      execAssignments xs (count+1)
-    parseAssignment :: SourceName -> String -> Either ParseError (Pattern, Expr)
-    parseAssignment = parse assignment
-      where
-        assignment :: GenParser Char st (Pattern, Expr)
-        assignment = do
-          key <- patternMatcher
-          _ <- matchTok '='
-          expr <- expr0
-          pure (key, expr)
+    opts = ScadOpts False False
+    show' = showErrorMessages "or" "unknown parse error" "expecting" "unexpected" "end of input" . errorMessages
+    execAssignments :: [String] -> StateC Fastℕ
+    execAssignments = foldlM execAssignment 0
```

Comment:
foldlM is doing the same job as the recursion, except that we can now lean on the base libraries to do the work for us.
foldlM is the same as foldl, except that it also carries the monadic effects through the computation.